src/lib/helpers.ts

```diff
@@ -194,23 +168,39 @@ export class Deferred<T = any> {
   }
 }
 
-// Taken from: https://stackoverflow.com/questions/38552003/how-to-decode-jwt-token-in-javascript-without-using-a-library
-export function decodeJWTPayload(token: string) {
-  // Regex checks for base64url format
-  const base64UrlRegex = /^([a-z0-9_-]{4})*($|[a-z0-9_-]{3}=?$|[a-z0-9_-]{2}(==)?$)$/i
-
+export function decodeJWT(token: string): {
+  header: JwtHeader
+  payload: JwtPayload
+  signature: Uint8Array
+  raw: {
+    header: string
+    payload: string
+  }
+} {
   const parts = token.split('.')
 
   if (parts.length !== 3) {
-    throw new Error('JWT is not valid: not a JWT structure')
+    throw new AuthInvalidJwtError('Invalid JWT structure')
   }
 
-  if (!base64UrlRegex.test(parts[1])) {
-    throw new Error('JWT is not valid: payload is not in base64url format')
+  // Regex checks for base64url format
+  const base64UrlRegex = /^([a-z0-9_-]{4})*($|[a-z0-9_-]{3}=?$|[a-z0-9_-]{2}(==)?$)$/i
+  for (let i = 0; i < parts.length; i++) {
+    if (!base64UrlRegex.test(parts[i])) {
+      throw new AuthInvalidJwtError('JWT not in base64url format')
+    }
   }
-
-  const base64Url = parts[1]
-  return JSON.parse(decodeBase64URL(base64Url))
+  const decoder = new TextDecoder()
+  const data = {
+    header: JSON.parse(decoder.decode(base64url.parse(parts[0], { loose: true }))),
+    payload: JSON.parse(decoder.decode(base64url.parse(parts[1], { loose: true }))),
+    signature: base64url.parse(parts[2], { loose: true }),
```

Comment:
Please use our library, it's been tested to support UTF-8 unlike other libraries / atob / btoa.